RakeshRenzous/space-rockets
# feat: Add favourites features to the app

**Description**

Adding a feature to the app to allow the users to mark their favourite items by respective group and to view them later.

- User can mark an item as a favourite from the list and details page for all items
- User can view their list of favourites in a slide-in drawer by clicking the button in the header and can navigate to their favourite items from there.
- User can remove items from list in the drawer as well from the list/details page of the item.
- The marked items will be persisted even after the app is closed.

**Steps to test**

- Visit home page and click on "Browse SpaceX Launches"
- Click on the heart icon of any launch item.
- Click on the "Favourites" button in the header - Should see the marked items.
- Click on item in the list - Should take to the details page.
- Should be able to see the item is already marked as favourite.
- Unmark the item by clicking on the heart icon.
- Click on the "Favourites" button in the header - Should see an empty message and item would have been removed.
- Add another item from the list page.
- Click on the "Favourites" button in the header and click on the heart icon of the item in the list - The item should be removed from the list.

**Thought process**

While working on this feature have tried restructure and refactor areas of code that were necessary for this feature. For the state management I explored options such as using redux, but since the current scope is that huge, felt using context would be more than sufficient for now and also reduces the overhead of another dependency. In case of future use cases which would require some more advanced tooling we can migrate to redux. For state persistence have used localStorage which I am syncing after every action. I have also placed the Favourites button in the header so that it can be accessed quickly from any page in the app.

**Favourites button**
<img width="1440" alt="Screenshot 2021-11-17 at 10 46 30 PM" src="https://user-images.githubusercontent.com/23741529/142253894-a1ead525-163a-4fa9-a64d-774d60084440.png">

**Favourites drawer**
<img width="1439" alt="Screenshot 2021-11-17 at 10 46 56 PM" src="https://user-images.githubusercontent.com/23741529/142253917-f4792216-8032-4f1c-aab6-19e6c7c923b5.png">
